profiling/src/io.rs

```diff
@@ -0,0 +1,683 @@
+
+            // Access the symbol name via the string table
+            let name_offset = (*sym).st_name as isize;
+            let name_ptr = strtab.offset(name_offset);
+            let name = CStr::from_ptr(name_ptr).to_str().unwrap_or("");
+
+            if name == overwrite.symbol_name {
+                // Calculate the GOT entry address
+                let got_entry =
+                    ((*info).dlpi_addr as usize + (*rel).r_offset as usize) as *mut *mut ();
+
+                // Change memory protection so we can write to the GOT entry
+                let page_size = libc::sysconf(libc::_SC_PAGESIZE) as usize;
+                let aligned_addr = (got_entry as usize) & !(page_size - 1);
+                if libc::mprotect(
+                    aligned_addr as *mut c_void,
+                    page_size,
+                    libc::PROT_READ | libc::PROT_WRITE,
+                ) != 0
+                {
+                    let err = *libc::__errno_location();
+                    trace!("mprotect failed: {}", err);
+                    return;
+                }
+
+                trace!(
+                    "Overriding GOT entry for {} at offset {:?} (abs: {:p}) pointing to {:p} (orig function at {:p})",
+                    overwrite.symbol_name,
+                    (*rel).r_offset,
+                    got_entry,
+                    *got_entry,
+                    *overwrite.orig_func
+                );
+
+                // This works for musl based linux distros, but not for libc once
+                *overwrite.orig_func = libc::dlsym(libc::RTLD_NEXT, name_ptr) as *mut ();
+                if (*overwrite.orig_func).is_null() {
+                    // libc linux fallback
+                    *overwrite.orig_func = *got_entry;
+                }
+                *got_entry = overwrite.new_func;
+                continue;
+            }
+        }
+    }
+}
+
+/// Callback function that should be passed to `libc::dl_iterate_phdr()` and gets called for every
+/// shared object.
+unsafe extern "C" fn callback(info: *mut dl_phdr_info, _size: usize, data: *mut c_void) -> c_int {
+    let overwrites = &mut *(data as *mut Vec<GotSymbolOverwrite>);
+
+    // detect myself ...
+    let mut my_info: libc::Dl_info = std::mem::zeroed();
+    if libc::dladdr(callback as *const c_void, &mut my_info) == 0 {
+        error!("Did not find my own `dladdr` and therefore can't hook into the GOT.");
+        return 0;
+    }
+    let my_base_addr = my_info.dli_fbase as usize;
+    let module_base_addr = (*info).dlpi_addr as usize;
+    if module_base_addr == my_base_addr {
+        // "this" lib is actually me: skipping GOT hooking for myself
+        return 0;
+    }
+
+    let name = if (*info).dlpi_name.is_null() || *(*info).dlpi_name == 0 {
+        "[Executable]"
+    } else {
+        CStr::from_ptr((*info).dlpi_name)
+            .to_str()
+            .unwrap_or("[Unknown]")
+    };
+
+    //libc::raise(libc::SIGTRAP);
```

Comment:
```suggestion
```
leftover comment?

Reply:
Jup, leftover from debugging 😉 
removed in commit 853e15b6f